File: src/particles/PythonComputer.cpp
#include "PythonComputer.hpp"
#include <boost/python.hpp>

using namespace boost::python;
using namespace espresso::particles;

void PythonComputer::bind(const Storage *storage) {
  particleId = storage-> getIdPropertyHandle();
}

void PythonComputer::operator()(const ParticleHandle pref) {
  get_override("each")(particleId[pref]);
}

void PythonComputer::registerPython() {
  class_<PythonComputer, bases<Computer> >
    ("particles_PythonComputer", init<>());
}

File: src/particles/PythonComputer.py
from espresso import pmi
from _espresso import particles_PythonComputer as _PythonComputer

__all__ = [ "PythonComputerLocal"]

class PythonComputerLocal(_PythonComputer) :
    """
    a per-particle-computer using Python code. You mainly need to
    provide two functions: "each", which is called for each particle
    and is mandatory to override, and "collect", which you only need
    to provide if you want something to happen after the function
    "each" has been called for each particle. The result of "collect"
    on the master node will be available for further processing, see
    the documentation of decomposition.Decomposer.foreach for further
    details.
    """
    def __init__(self):
        _PythonComputer.__init__(self)

    def __apply__(self, id):
        """
        override this function to implement a per-particle computer. This routine will be
        called for each particle, with id being the identity of the particle.
        """
        raise RuntimeError("PythonComputer invoked without __apply__ being implemented.")
